packages/vite/src/node/server/environments/ssrEnvironment.ts

```diff
@@ -0,0 +1,85 @@
+import path from 'node:path'
+import fs from 'node:fs'
+import { DevEnvironment } from '../environment'
+import type { ServerHMRChannel } from '../hmr'
+import type { ViteDevServer } from '../index'
+import { asyncFunctionDeclarationPaddingLineCount } from '../../../shared/utils'
+
+export function createSsrEnvironment(
+  hotChannel: ServerHMRChannel,
+  server: ViteDevServer,
+  name: string,
+): DevEnvironment {
+  const environment = new DevEnvironment(server, name, {
+    hot: hotChannel,
+    ssr: {
+      async processRequest(req, res) {
+        const url = req.url
+
+        assert(url, () => `Request URL is missing.`)
+
+        const entry = server.config.ssr.entry
+
+        assert(entry, () => `SSR entry is not specified.`)
+
+        let template = fs.readFileSync(
+          path.resolve(server.config.root, 'index.html'),
+          'utf-8',
+        )
+
+        template = await server.transformIndexHtml(url, template)
+
+        assert(
+          template.includes(`<!--ssr-outlet-->`),
+          () => `index.html is missing the "<!--ssr-outlet-->" comment.`,
+        )
+
+        const resolvedEntry = await server.pluginContainer.resolveId(
+          entry,
+          undefined,
+          { environment },
+        )
+
+        assert(resolvedEntry, () => `Failed to resolve SSR entry: ${entry}`)
+
+        const { render } = await server.nodeModuleRunner.import(
+          resolvedEntry.id,
+        )
```

Comment:
This is interesting. It may be too much on the convention side. Not only `render` but also `index.html`. I'm thinking that in the same way the user will need to write the workerd server entry (including importing the HTML template, and calling `import` on the runner themselves), maybe it is ok for the user to also do the same in the default node runner. I think we are missing another abstraction here. `environment.ssr.processRequest(req, res)` in workerd would end up dispatching a fetch inside workerd. Maybe we should think that the node version should do the same, ending up calling the the current code in this function provided by the user somehow 🤔